File: sdk/typescript/src/evaluator.ts
import type { AgentMandate, TransactionEnvelope } from "./index.js"

export interface UsageSnapshot {
  calls?: number
  amount?: string
  currency?: string
  quantity?: string
  consumed?: boolean
  idempotency_keys?: string[]
  minted_supply?: string
}

export interface Consumption {
  calls?: number
  amount?: string
  currency?: string
  quantity?: string
  idempotency_key?: string
  consume?: boolean
}

export interface CommerceEvaluationContext {
  merchant_organisation_id: string
  service_id: string
  offer_id: string
  currency: string
  quantity: number
  unit_price: string
  total_amount: string
  idempotency_key: string
  terms_digest?: string
}

export interface RwaEvaluationContext {
  asset_id: string
  state_claim_id: string
  network: string
  token_contract: string
  operation: "mint" | "burn" | "transfer" | "redeem" | "publish_state"
  unit: string
  quantity: string
  reserve: string
  approval_count: number
  approval_roles: string[]
  current_supply: string
  maximum_supply?: string
  claim_valid_until: string
}

export interface EvaluationRequest {
  oati_version: "1.0"
  evaluation_time: string
  mandate: AgentMandate
  parent_mandate?: AgentMandate
  delegation_depth?: number
  envelope: TransactionEnvelope
  usage: UsageSnapshot
  consumption?: Consumption
  commerce?: CommerceEvaluationContext
  rwa?: RwaEvaluationContext
}

export interface EvaluationResult {
  oati_version: "1.0"
  decision: "allow" | "deny"
  mandate_id: string
  transaction_id: string
  reason_codes: string[]
  next_usage: UsageSnapshot
}

/** Deterministically evaluate authority from explicit objects, time, and prior usage. */
export function evaluateAuthority(request: EvaluationRequest): EvaluationResult {
  const reasons = new Set<string>()
  const now = timestamp(request.evaluation_time)
  const mandate = request.mandate
  const envelope = request.envelope
  if (mandate.status !== "active") reasons.add("MANDATE_NOT_ACTIVE")
  if (timestamp(mandate.not_before) > now) reasons.add("MANDATE_NOT_YET_ACTIVE")
  if (timestamp(mandate.expires_at) <= now) reasons.add("MANDATE_EXPIRED")
  if (envelope.mandate_id !== mandate.id) reasons.add("MANDATE_REFERENCE_MISMATCH")
  if (envelope.agent_id !== mandate.subject) reasons.add("SUBJECT_MISMATCH")
  if (!mandate.actions.includes(envelope.action)) reasons.add("ACTION_NOT_ALLOWED")
  checkSetConstraint(mandate.resources, envelope.resource, "RESOURCE_NOT_ALLOWED", reasons)
  if (envelope.purpose !== mandate.purpose) reasons.add("PURPOSE_MISMATCH")
  checkSetConstraint(mandate.counterparties, envelope.counterparty, "COUNTERPARTY_NOT_ALLOWED", reasons)
  checkSetConstraint(mandate.destinations, envelope.destination, "DESTINATION_NOT_ALLOWED", reasons)
  if (envelope.profile !== undefined && envelope.profile !== mandate.profile) reasons.add("PROFILE_MISMATCH")
  if (request.parent_mandate) checkChildMandate(mandate, request.parent_mandate, request.delegation_depth ?? 1, now, reasons)
  else if (mandate.parent_mandate) reasons.add("PARENT_MANDATE_REQUIRED")

  const usage = normalizedUsage(request.usage)
  const delta = effectiveConsumption(request)
  checkConsumptionContext(request, delta, reasons)
  checkConsumption(mandate, usage, delta, reasons)
  if (request.commerce || mandate.profile === "https://specs.intelliger.ai/oati/profiles/commerce/v0.1") checkCommerce(mandate, envelope, request.commerce, usage, reasons)
  if (request.rwa || mandate.profile === "https://specs.intelliger.ai/oati/profiles/rwa/v0.1") checkRwa(mandate, envelope, request.rwa, usage, request.usage.minted_supply !== undefined, now, reasons)

  const reasonCodes = [...reasons].sort()
  return {
    oati_version: "1.0",
    decision: reasonCodes.length === 0 ? "allow" : "deny",
    mandate_id: mandate.id,
    transaction_id: envelope.id,
    reason_codes: reasonCodes,
    next_usage: reasonCodes.length === 0 ? applyConsumption(usage, delta, request) : usage,
  }
}

function checkChildMandate(child: AgentMandate, parent: AgentMandate, depth: number, now: number, reasons: Set<string>): void {
  if (child.parent_mandate !== parent.id) reasons.add("PARENT_MANDATE_MISMATCH")
  if (parent.status !== "active" || timestamp(parent.not_before) > now || timestamp(parent.expires_at) <= now) reasons.add("PARENT_MANDATE_NOT_ACTIVE")
  if (!parent.delegation?.allowed || depth > parent.delegation.max_depth) reasons.add("DELEGATION_NOT_ALLOWED")
  if (!subset(child.actions, parent.actions)) reasons.add("CHILD_ACTION_AMPLIFICATION")
  if (!constrainedSubset(child.resources, parent.resources)) reasons.add("CHILD_RESOURCE_AMPLIFICATION")
  if (!constrainedSubset(child.counterparties, parent.counterparties)) reasons.add("CHILD_COUNTERPARTY_AMPLIFICATION")
  if (!constrainedSubset(child.destinations, parent.destinations)) reasons.add("CHILD_DESTINATION_AMPLIFICATION")
  if (child.purpose !== parent.purpose) reasons.add("CHILD_PURPOSE_AMPLIFICATION")
  if (timestamp(child.not_before) < timestamp(parent.not_before) || timestamp(child.expires_at) > timestamp(parent.expires_at)) reasons.add("CHILD_TIME_AMPLIFICATION")
  if (!narrowerObject(child.limits, parent.limits)) reasons.add("CHILD_LIMIT_AMPLIFICATION")
  if (!narrowerObject(child.data_use, parent.data_use)) reasons.add("CHILD_DATA_USE_AMPLIFICATION")
  if (child.delegation?.allowed) {
    const remaining = Math.max(0, parent.delegation!.max_depth - depth)
    if (child.delegation.max_depth > remaining) reasons.add("CHILD_DELEGATION_AMPLIFICATION")
  }
  checkProfileSubset(child, parent, reasons)
}

function checkProfileSubset(child: AgentMandate, parent: AgentMandate, reasons: Set<string>): void {
  if (child.profile !== parent.profile) { if (child.profile || parent.profile) reasons.add("CHILD_PROFILE_AMPLIFICATION"); return }
  const childCommerce = objectAt(child.extensions, "commerce"), parentCommerce = objectAt(parent.extensions, "commerce")
  if (parentCommerce && (!childCommerce || !sameFields(childCommerce, parentCommerce, ["merchant_organisation_id", "service_id", "offer_id", "currency", "billing_model", "terms_digest"])
    || !decimalAtMost(childCommerce.max_unit_price, parentCommerce.max_unit_price) || !decimalAtMost(childCommerce.max_total, parentCommerce.max_total)
    || !numberAtMost(childCommerce.max_quantity, parentCommerce.max_quantity))) reasons.add("CHILD_COMMERCE_AMPLIFICATION")
  const childRwa = objectAt(child.extensions, "rwa"), parentRwa = objectAt(parent.extensions, "rwa")
  if (parentRwa) {
    const requiredParent = stringArray(parentRwa.required_roles), requiredChild = stringArray(childRwa?.required_roles)
    if (!childRwa || !sameFields(childRwa, parentRwa, ["asset_id", "state_claim_id", "network", "token_contract", "operation", "unit"])
      || !decimalAtMost(childRwa.max_quantity, parentRwa.max_quantity) || !numberAtLeast(childRwa.minimum_approvals, parentRwa.minimum_approvals)
      || !subset(requiredParent, requiredChild) || parentRwa.one_time === true && childRwa.one_time !== true) reasons.add("CHILD_RWA_AMPLIFICATION")
  }
}

function checkConsumption(mandate: AgentMandate, usage: Required<UsageSnapshot>, delta: Required<Consumption>, reasons: Set<string>): void {
  if (usage.consumed) reasons.add("MANDATE_ALREADY_CONSUMED")
  if (delta.idempotency_key && usage.idempotency_keys.includes(delta.idempotency_key)) reasons.add("IDEMPOTENCY_REPLAY")
  const limits = mandate.limits ?? {}
  if (typeof limits.max_calls === "number" && usage.calls + delta.calls > limits.max_calls) reasons.add("CALL_LIMIT_EXCEEDED")
  if (mandate.profile === undefined && (typeof limits.max_quantity === "string" || typeof limits.max_quantity === "number")
    && compareDecimal(addDecimal(usage.quantity, delta.quantity), String(limits.max_quantity)) > 0) reasons.add("QUANTITY_LIMIT_EXCEEDED")
  if (typeof limits.max_total === "string" && delta.amount !== "0") {
    const currency = typeof limits.currency === "string" ? limits.currency : undefined
    if (currency && (delta.currency !== currency || usage.amount !== "0" && usage.currency !== currency)) reasons.add("BUDGET_CURRENCY_MISMATCH")
    if (compareDecimal(addDecimal(usage.amount, delta.amount), limits.max_total) > 0) reasons.add("BUDGET_EXCEEDED")
  }
}

function checkCommerce(mandate: AgentMandate, envelope: TransactionEnvelope, context: CommerceEvaluationContext | undefined, usage: Required<UsageSnapshot>, reasons: Set<string>): void {
  const limits = objectAt(mandate.extensions, "commerce")
  if (!context || !limits) { reasons.add("COMMERCE_CONTEXT_REQUIRED"); return }
  const signed = objectAt(envelope.extensions, "commerce")
  const envelopeMismatch = signed && (envelope.resource !== context.service_id || envelope.counterparty !== context.merchant_organisation_id
    || !sameMappedFields(signed, context as unknown as Record<string, unknown>, [
      ["offer_id", "offer_id"], ["currency", "currency"], ["quantity", "quantity"], ["quoted_unit_price", "unit_price"],
      ["quoted_total", "total_amount"], ["idempotency_key", "idempotency_key"], ["terms_digest", "terms_digest"],
    ]))
  if (context.merchant_organisation_id !== limits.merchant_organisation_id) reasons.add("COMMERCE_MERCHANT_NOT_ALLOWED")
  if (context.service_id !== limits.service_id) reasons.add("COMMERCE_SERVICE_NOT_ALLOWED")
  if (context.offer_id !== limits.offer_id) reasons.add("COMMERCE_OFFER_NOT_ALLOWED")
  if (context.currency !== limits.currency || usage.amount !== "0" && usage.currency !== context.currency) reasons.add("COMMERCE_CURRENCY_MISMATCH")
  if (compareDecimal(context.unit_price, String(limits.max_unit_price)) > 0) reasons.add("COMMERCE_UNIT_PRICE_EXCEEDED")
  if (compareDecimal(context.total_amount, multiplyDecimal(context.unit_price, context.quantity)) !== 0) reasons.add("COMMERCE_TOTAL_INVALID")
  if (compareDecimal(addDecimal(usage.amount, context.total_amount), String(limits.max_total)) > 0) reasons.add("COMMERCE_BUDGET_EXCEEDED")
  if (compareDecimal(String(context.quantity), String(limits.max_quantity)) > 0) reasons.add("COMMERCE_QUANTITY_EXCEEDED")
  if (limits.terms_digest && context.terms_digest !== limits.terms_digest) reasons.add("COMMERCE_TERMS_MISMATCH")
  if (usage.idempotency_keys.includes(context.idempotency_key)) reasons.add("IDEMPOTENCY_REPLAY")
  // Preserve the established primary domain failure when an already-invalid
  // context also differs from the Envelope; otherwise report substitution.
  if (envelopeMismatch && ![...reasons].some((code) => code.startsWith("COMMERCE_"))) reasons.add("COMMERCE_ENVELOPE_CONTEXT_MISMATCH")
}

function checkRwa(mandate: AgentMandate, envelope: TransactionEnvelope, context: RwaEvaluationContext | undefined, usage: Required<UsageSnapshot>, hasMintedSupply: boolean, now: number, reasons: Set<string>): void {
  const limits = objectAt(mandate.extensions, "rwa")
  if (!context || !limits) { reasons.add("RWA_CONTEXT_REQUIRED"); return }
  const signed = objectAt(envelope.extensions, "rwa")
  if (signed && (envelope.resource !== context.asset_id || !sameFields(signed, context as unknown as Record<string, unknown>,
    ["asset_id", "state_claim_id", "network", "token_contract", "operation", "unit", "quantity"]))) reasons.add("RWA_ENVELOPE_CONTEXT_MISMATCH")
  if (!sameFields(context as unknown as Record<string, unknown>, limits, ["asset_id", "state_claim_id", "network", "token_contract", "operation", "unit"])) reasons.add("RWA_TARGET_MISMATCH")
  if (timestamp(context.claim_valid_until) <= now) reasons.add("RWA_STATE_CLAIM_EXPIRED")
  if (compareDecimal(addDecimal(usage.quantity, context.quantity), String(limits.max_quantity)) > 0) reasons.add("RWA_QUANTITY_EXCEEDED")
  const resultingSupply = addDecimal(context.current_supply, context.quantity)
  if (compareDecimal(resultingSupply, context.reserve) > 0) reasons.add("RWA_RESERVE_EXCEEDED")
  if (context.maximum_supply && compareDecimal(resultingSupply, context.maximum_supply) > 0) reasons.add("RWA_MAXIMUM_SUPPLY_EXCEEDED")
  if (hasMintedSupply && compareDecimal(usage.minted_supply, context.current_supply) !== 0) reasons.add("RWA_SUPPLY_STATE_MISMATCH")
  if (context.approval_count < Number(limits.minimum_approvals)) reasons.add("RWA_APPROVAL_THRESHOLD_NOT_MET")
  for (const role of stringArray(limits.required_roles)) if (!context.approval_roles.includes(role)) reasons.add("RWA_REQUIRED_ROLE_MISSING")
  if (limits.one_time === true && usage.consumed) reasons.add("MANDATE_ALREADY_CONSUMED")
}

function effectiveConsumption(request: EvaluationRequest): Required<Consumption> {
  const commerce = request.commerce, rwa = request.rwa, supplied = request.consumption ?? {}
  return {
    calls: supplied.calls ?? 1,
    amount: commerce?.total_amount ?? supplied.amount ?? "0",
    currency: commerce?.currency ?? supplied.currency ?? "",
    quantity: commerce ? String(commerce.quantity) : rwa?.quantity ?? supplied.quantity ?? "0",
    idempotency_key: commerce?.idempotency_key ?? supplied.idempotency_key ?? "",
    consume: Boolean(objectAt(request.mandate.extensions, "rwa")?.one_time || request.mandate.limits?.one_time) || supplied.consume === true,
  }
}

function checkConsumptionContext(request: EvaluationRequest, effective: Required<Consumption>, reasons: Set<string>): void {
  const supplied = request.consumption
  if (!supplied) return
  const constrained: Array<keyof Consumption> = request.commerce
    ? ["amount", "currency", "quantity", "idempotency_key"]
    : request.rwa ? ["quantity"] : []
  if (effective.consume && supplied.consume === false) constrained.push("consume")
  if (constrained.some((field) => supplied[field] !== undefined && supplied[field] !== effective[field])) reasons.add("CONSUMPTION_CONTEXT_MISMATCH")
}

function applyConsumption(usage: Required<UsageSnapshot>, delta: Required<Consumption>, request: EvaluationRequest): UsageSnapshot {
  const next: UsageSnapshot = {
    calls: usage.calls + delta.calls,
    amount: addDecimal(usage.amount, delta.amount),
    currency: delta.currency || usage.currency,
    quantity: addDecimal(usage.quantity, delta.quantity),
    consumed: usage.consumed || delta.consume,
    idempotency_keys: delta.idempotency_key ? [...usage.idempotency_keys, delta.idempotency_key].sort() : [...usage.idempotency_keys],
    minted_supply: request.rwa ? addDecimal(request.rwa.current_supply, request.rwa.quantity) : usage.minted_supply,
  }
  return next
}

function normalizedUsage(value: UsageSnapshot): Required<UsageSnapshot> { return { calls: value.calls ?? 0, amount: value.amount ?? "0", currency: value.currency ?? "", quantity: value.quantity ?? "0", consumed: value.consumed ?? false, idempotency_keys: [...(value.idempotency_keys ?? [])].sort(), minted_supply: value.minted_supply ?? "0" } }
function checkSetConstraint(allowed: string[] | undefined, actual: string | undefined, code: string, reasons: Set<string>): void { if (allowed && (!actual || !allowed.includes(actual))) reasons.add(code) }
function constrainedSubset(child: string[] | undefined, parent: string[] | undefined): boolean { if (!parent) return true; return child !== undefined && subset(child, parent) }
function subset(child: string[], parent: string[]): boolean { return child.every((item) => parent.includes(item)) }
function timestamp(value: string): number { const parsed = Date.parse(value); if (Number.isNaN(parsed)) throw new TypeError(`Invalid RFC 3339 timestamp: ${value}`); return parsed }
function objectAt(value: Record<string, unknown> | undefined, key: string): Record<string, unknown> | undefined { const item = value?.[key]; return typeof item === "object" && item !== null && !Array.isArray(item) ? item as Record<string, unknown> : undefined }
function stringArray(value: unknown): string[] { return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [] }
function sameFields(child: Record<string, unknown>, parent: Record<string, unknown>, fields: string[]): boolean { return fields.every((field) => child[field] === parent[field]) }
function sameMappedFields(left: Record<string, unknown>, right: Record<string, unknown>, fields: Array<readonly [string, string]>): boolean { return fields.every(([leftField, rightField]) => left[leftField] === right[rightField]) }
function numberAtMost(child: unknown, parent: unknown): boolean { return typeof child === "number" && typeof parent === "number" && child <= parent }
function numberAtLeast(child: unknown, parent: unknown): boolean { return typeof child === "number" && typeof parent === "number" && child >= parent }
function decimalAtMost(child: unknown, parent: unknown): boolean { return typeof child === "string" && typeof parent === "string" && compareDecimal(child, parent) <= 0 }

function narrowerObject(child: Record<string, unknown> | undefined, parent: Record<string, unknown> | undefined): boolean {
  if (!parent) return true
  if (!child) return false
  return Object.entries(parent).every(([key, parentValue]) => {
    const childValue = child[key]
    if (typeof parentValue === "number") return typeof childValue === "number" && childValue <= parentValue
    if (typeof parentValue === "string" && isDecimal(parentValue)) return typeof childValue === "string" && isDecimal(childValue) && compareDecimal(childValue, parentValue) <= 0
    if (Array.isArray(parentValue)) return Array.isArray(childValue) && childValue.every((item) => parentValue.includes(item))
    if (typeof parentValue === "object" && parentValue !== null) return typeof childValue === "object" && childValue !== null && !Array.isArray(childValue) && narrowerObject(childValue as Record<string, unknown>, parentValue as Record<string, unknown>)
    return childValue === parentValue
  })
}

interface Decimal { coefficient: bigint; scale: number }
function parseDecimal(value: string): Decimal { if (!isDecimal(value)) throw new TypeError(`Invalid decimal: ${value}`); const [whole, fraction = ""] = value.split("."); return { coefficient: BigInt(whole + fraction), scale: fraction.length } }
function align(left: Decimal, right: Decimal): [bigint, bigint, number] { const scale = Math.max(left.scale, right.scale); return [left.coefficient * 10n ** BigInt(scale - left.scale), right.coefficient * 10n ** BigInt(scale - right.scale), scale] }
function compareDecimal(left: string, right: string): number { const [a, b] = align(parseDecimal(left), parseDecimal(right)); return a === b ? 0 : a > b ? 1 : -1 }
function addDecimal(left: string, right: string): string { const [a, b, scale] = align(parseDecimal(left), parseDecimal(right)); return formatDecimal(a + b, scale) }
function multiplyDecimal(value: string, multiplier: number): string { const parsed = parseDecimal(value); return formatDecimal(parsed.coefficient * BigInt(multiplier), parsed.scale) }
function formatDecimal(coefficient: bigint, scale: number): string { if (scale === 0) return coefficient.toString(); const digits = coefficient.toString().padStart(scale + 1, "0"); return `${digits.slice(0, -scale)}.${digits.slice(-scale)}` }
function isDecimal(value: string): boolean { return /^(0|[1-9][0-9]*)(\.[0-9]+)?$/.test(value) }

File: sdk/python/src/oati/evaluator.py
from __future__ import annotations
from copy import deepcopy
from datetime import datetime
from decimal import Decimal
from typing import Any

COMMERCE="https://specs.intelliger.ai/oati/profiles/commerce/v0.1"; RWA="https://specs.intelliger.ai/oati/profiles/rwa/v0.1"
def evaluate_authority(request:dict[str,Any])->dict[str,Any]:
    m=request["mandate"]; e=request["envelope"]; now=_time(request["evaluation_time"]); reasons:set[str]=set()
    if m.get("status")!="active": reasons.add("MANDATE_NOT_ACTIVE")
    if _time(m["not_before"])>now: reasons.add("MANDATE_NOT_YET_ACTIVE")
    if _time(m["expires_at"])<=now: reasons.add("MANDATE_EXPIRED")
    if e.get("mandate_id")!=m.get("id"): reasons.add("MANDATE_REFERENCE_MISMATCH")
    if e.get("agent_id")!=m.get("subject"): reasons.add("SUBJECT_MISMATCH")
    if e.get("action") not in m.get("actions",[]): reasons.add("ACTION_NOT_ALLOWED")
    _constraint(m,"resources",e.get("resource"),"RESOURCE_NOT_ALLOWED",reasons)
    if e.get("purpose")!=m.get("purpose"): reasons.add("PURPOSE_MISMATCH")
    _constraint(m,"counterparties",e.get("counterparty"),"COUNTERPARTY_NOT_ALLOWED",reasons); _constraint(m,"destinations",e.get("destination"),"DESTINATION_NOT_ALLOWED",reasons)
    if "profile" in e and e.get("profile")!=m.get("profile"): reasons.add("PROFILE_MISMATCH")
    if request.get("parent_mandate"): _child(m,request["parent_mandate"],request.get("delegation_depth",1),now,reasons)
    elif m.get("parent_mandate"): reasons.add("PARENT_MANDATE_REQUIRED")
    usage=_usage(request.get("usage",{})); delta=_consumption(request,m)
    _consumption_context(request,delta,reasons)
    _limits(m,usage,delta,reasons)
    if request.get("commerce") is not None or m.get("profile")==COMMERCE: _commerce(m,e,request.get("commerce"),usage,reasons)
    if request.get("rwa") is not None or m.get("profile")==RWA: _rwa(m,e,request.get("rwa"),usage,"minted_supply" in request.get("usage",{}),now,reasons)
    codes=sorted(reasons)
    return {"oati_version":"1.0","decision":"allow" if not codes else "deny","mandate_id":m["id"],"transaction_id":e["id"],"reason_codes":codes,"next_usage":_apply(usage,delta,request.get("rwa")) if not codes else usage}

def _child(c,p,depth,now,r):
    if c.get("parent_mandate")!=p.get("id"):r.add("PARENT_MANDATE_MISMATCH")
    if p.get("status")!="active" or _time(p["not_before"])>now or _time(p["expires_at"])<=now:r.add("PARENT_MANDATE_NOT_ACTIVE")
    d=p.get("delegation",{});
    if not d.get("allowed") or depth>d.get("max_depth",0):r.add("DELEGATION_NOT_ALLOWED")
    if not _subset(c.get("actions",[]),p.get("actions",[])):r.add("CHILD_ACTION_AMPLIFICATION")
    for field,code in (("resources","CHILD_RESOURCE_AMPLIFICATION"),("counterparties","CHILD_COUNTERPARTY_AMPLIFICATION"),("destinations","CHILD_DESTINATION_AMPLIFICATION")):
        if field in p and (field not in c or not _subset(c[field],p[field])):r.add(code)
    if c.get("purpose")!=p.get("purpose"):r.add("CHILD_PURPOSE_AMPLIFICATION")
    if _time(c["not_before"])<_time(p["not_before"]) or _time(c["expires_at"])>_time(p["expires_at"]):r.add("CHILD_TIME_AMPLIFICATION")
    if not _narrow(c.get("limits"),p.get("limits")):r.add("CHILD_LIMIT_AMPLIFICATION")
    if not _narrow(c.get("data_use"),p.get("data_use")):r.add("CHILD_DATA_USE_AMPLIFICATION")
    cd=c.get("delegation");
    if cd and cd.get("allowed") and cd.get("max_depth",0)>max(0,d.get("max_depth",0)-depth):r.add("CHILD_DELEGATION_AMPLIFICATION")
    if c.get("profile")!=p.get("profile"):
        if c.get("profile") or p.get("profile"):r.add("CHILD_PROFILE_AMPLIFICATION")
        return
    ce=c.get("extensions",{});pe=p.get("extensions",{});cc=ce.get("commerce");pc=pe.get("commerce")
    if pc and (not cc or not _same(cc,pc,("merchant_organisation_id","service_id","offer_id","currency","billing_model","terms_digest")) or Decimal(cc["max_unit_price"])>Decimal(pc["max_unit_price"]) or Decimal(cc["max_total"])>Decimal(pc["max_total"]) or cc["max_quantity"]>pc["max_quantity"]):r.add("CHILD_COMMERCE_AMPLIFICATION")
    cr=ce.get("rwa");pr=pe.get("rwa")
    if pr and (not cr or not _same(cr,pr,("asset_id","state_claim_id","network","token_contract","operation","unit")) or Decimal(cr["max_quantity"])>Decimal(pr["max_quantity"]) or cr["minimum_approvals"]<pr["minimum_approvals"] or not _subset(pr.get("required_roles",[]),cr.get("required_roles",[])) or pr.get("one_time") is True and cr.get("one_time") is not True):r.add("CHILD_RWA_AMPLIFICATION")

def _limits(m,u,d,r):
    if u["consumed"]:r.add("MANDATE_ALREADY_CONSUMED")
    if d["idempotency_key"] and d["idempotency_key"] in u["idempotency_keys"]:r.add("IDEMPOTENCY_REPLAY")
    limits=m.get("limits",{})
    if "max_calls" in limits and u["calls"]+d["calls"]>limits["max_calls"]:r.add("CALL_LIMIT_EXCEEDED")
    if m.get("profile") is None and "max_quantity" in limits and Decimal(u["quantity"])+Decimal(d["quantity"])>Decimal(str(limits["max_quantity"])):r.add("QUANTITY_LIMIT_EXCEEDED")
    if "max_total" in limits and d["amount"]!="0":
        currency=limits.get("currency")
        if currency and (d["currency"]!=currency or u["amount"]!="0" and u["currency"]!=currency):r.add("BUDGET_CURRENCY_MISMATCH")
        if Decimal(u["amount"])+Decimal(d["amount"])>Decimal(limits["max_total"]):r.add("BUDGET_EXCEEDED")
def _commerce(m,e,c,u,r):
    limits=m.get("extensions",{}).get("commerce")
    if not c or not limits:r.add("COMMERCE_CONTEXT_REQUIRED");return
    signed=e.get("extensions",{}).get("commerce")
    envelope_mismatch=bool(signed) and (e.get("resource")!=c.get("service_id") or e.get("counterparty")!=c.get("merchant_organisation_id") or not _mapped(signed,c,(("offer_id","offer_id"),("currency","currency"),("quantity","quantity"),("quoted_unit_price","unit_price"),("quoted_total","total_amount"),("idempotency_key","idempotency_key"),("terms_digest","terms_digest"))))
    for f,code in (("merchant_organisation_id","COMMERCE_MERCHANT_NOT_ALLOWED"),("service_id","COMMERCE_SERVICE_NOT_ALLOWED"),("offer_id","COMMERCE_OFFER_NOT_ALLOWED")):
        if c.get(f)!=limits.get(f):r.add(code)
    if c["currency"]!=limits["currency"] or u["amount"]!="0" and u["currency"]!=c["currency"]:r.add("COMMERCE_CURRENCY_MISMATCH")
    if Decimal(c["unit_price"])>Decimal(limits["max_unit_price"]):r.add("COMMERCE_UNIT_PRICE_EXCEEDED")
    if Decimal(c["total_amount"])!=Decimal(c["unit_price"])*c["quantity"]:r.add("COMMERCE_TOTAL_INVALID")
    if Decimal(u["amount"])+Decimal(c["total_amount"])>Decimal(limits["max_total"]):r.add("COMMERCE_BUDGET_EXCEEDED")
    if Decimal(c["quantity"])>Decimal(limits["max_quantity"]):r.add("COMMERCE_QUANTITY_EXCEEDED")
    if limits.get("terms_digest") and c.get("terms_digest")!=limits["terms_digest"]:r.add("COMMERCE_TERMS_MISMATCH")
    if c["idempotency_key"] in u["idempotency_keys"]:r.add("IDEMPOTENCY_REPLAY")
    if envelope_mismatch and not any(code.startswith("COMMERCE_") for code in r):r.add("COMMERCE_ENVELOPE_CONTEXT_MISMATCH")
def _rwa(m,e,c,u,has_supply,now,r):
    limits=m.get("extensions",{}).get("rwa")
    if not c or not limits:r.add("RWA_CONTEXT_REQUIRED");return
    signed=e.get("extensions",{}).get("rwa")
    if signed and (e.get("resource")!=c.get("asset_id") or not _same(signed,c,("asset_id","state_claim_id","network","token_contract","operation","unit","quantity"))):r.add("RWA_ENVELOPE_CONTEXT_MISMATCH")
    if not _same(c,limits,("asset_id","state_claim_id","network","token_contract","operation","unit")):r.add("RWA_TARGET_MISMATCH")
    if _time(c["claim_valid_until"])<=now:r.add("RWA_STATE_CLAIM_EXPIRED")
    if Decimal(u["quantity"])+Decimal(c["quantity"])>Decimal(limits["max_quantity"]):r.add("RWA_QUANTITY_EXCEEDED")
    resulting=Decimal(c["current_supply"])+Decimal(c["quantity"])
    if resulting>Decimal(c["reserve"]):r.add("RWA_RESERVE_EXCEEDED")
    if c.get("maximum_supply") and resulting>Decimal(c["maximum_supply"]):r.add("RWA_MAXIMUM_SUPPLY_EXCEEDED")
    if has_supply and Decimal(u["minted_supply"])!=Decimal(c["current_supply"]):r.add("RWA_SUPPLY_STATE_MISMATCH")
    if c["approval_count"]<limits["minimum_approvals"]:r.add("RWA_APPROVAL_THRESHOLD_NOT_MET")
    if any(role not in c["approval_roles"] for role in limits.get("required_roles",[])):r.add("RWA_REQUIRED_ROLE_MISSING")
    if limits.get("one_time") and u["consumed"]:r.add("MANDATE_ALREADY_CONSUMED")
def _consumption(req,m):
    c=req.get("commerce");w=req.get("rwa");supplied=req.get("consumption",{})
    d={"calls":supplied.get("calls",1),"amount":supplied.get("amount","0"),"currency":supplied.get("currency",""),"quantity":supplied.get("quantity","0"),"idempotency_key":supplied.get("idempotency_key",""),"consume":supplied.get("consume") is True}
    if c:d.update(amount=c["total_amount"],currency=c["currency"],quantity=str(c["quantity"]),idempotency_key=c["idempotency_key"])
    elif w:d["quantity"]=w["quantity"]
    d["consume"]=d["consume"] or bool(m.get("extensions",{}).get("rwa",{}).get("one_time") or m.get("limits",{}).get("one_time"))
    return d
def _consumption_context(req,effective,r):
    supplied=req.get("consumption")
    if not supplied:return
    fields=("amount","currency","quantity","idempotency_key") if req.get("commerce") else ("quantity",) if req.get("rwa") else ()
    constrained=list(fields)
    if effective["consume"] and supplied.get("consume") is False:constrained.append("consume")
    if any(field in supplied and supplied[field]!=effective[field] for field in constrained):r.add("CONSUMPTION_CONTEXT_MISMATCH")
def _usage(v):return {"calls":v.get("calls",0),"amount":v.get("amount","0"),"currency":v.get("currency",""),"quantity":v.get("quantity","0"),"consumed":v.get("consumed",False),"idempotency_keys":sorted(v.get("idempotency_keys",[])),"minted_supply":v.get("minted_supply","0")}
def _apply(u,d,rwa):
    n=deepcopy(u);n.update(calls=u["calls"]+d["calls"],amount=_fmt(Decimal(u["amount"])+Decimal(d["amount"])),currency=d["currency"] or u["currency"],quantity=_fmt(Decimal(u["quantity"])+Decimal(d["quantity"])),consumed=u["consumed"] or d["consume"]);n["idempotency_keys"]=sorted(u["idempotency_keys"]+([d["idempotency_key"]] if d["idempotency_key"] else []));n["minted_supply"]=_fmt(Decimal(rwa["current_supply"])+Decimal(rwa["quantity"])) if rwa else u["minted_supply"];return n
def _narrow(c,p):
    if p is None:return True
    if c is None:return False
    for k,v in p.items():
        if k not in c:return False
        x=c[k]
        if isinstance(v,(int,float)) and x>v:return False
        if isinstance(v,str):
            try:
                if Decimal(x)>Decimal(v):return False
            except: 
                if x!=v:return False
        elif isinstance(v,list) and not _subset(x,v):return False
        elif isinstance(v,dict) and not _narrow(x,v):return False
        elif not isinstance(v,(str,list,dict,int,float)) and x!=v:return False
    return True
def _constraint(o,f,a,code,r):
    if f in o and a not in o[f]:r.add(code)
def _subset(a,b):return all(x in b for x in a)
def _same(a,b,fields):return all(a.get(f)==b.get(f) for f in fields)
def _mapped(a,b,fields):return all(a.get(left)==b.get(right) for left,right in fields)
def _time(v):return datetime.fromisoformat(v.replace("Z","+00:00"))
def _fmt(v:Decimal):return format(v,"f")
